### PR TITLE
Version 2.0

### PR DESCRIPTION
**Pull request by bencorlett**
# Version 2
- [x] Move to PSR-7 (Guzzle 6) #27 
- [x] Consistent API with OAuth 2 client #28

### CODE REVIEW

**Comment by bencorlett** [comment]
This PR replaces #34.

**Comment by ajibarra** [comment]
Hi @bencorlett. Are you planning to merge this changes soon to develop or master? I am looking for the standard between both oauth1 and oauth2 client. Great work btw.

**Comment by bencorlett** [comment]
@ajibarra soon(ish), still working through the rewrite for v2 in my spare time. No real ETA as yet unfortunately. v1 is stable, just has an API that doesn't match the OAuth 2 client (it used to, but OAuth 2 has since moved forward and I'm catching up here).

**Comment by stevenmaguire** [comment]
@bencorlett you're making things happen! I'm happy to help out here and I don't want to impede your existing flow. If you can provide a deeper layer of detail about your current plan/thought process for the refactor, I can take care of some of the work.

**Comment by stevenmaguire** [comment]
#40 encapsulates a bit of refactoring that I've been wanting to do. I've moved a handful of methods that are factories or static in nature to appropriate classes and created new classes where appropriate.

**Comment by stevenmaguire** [comment]
> Send arbitrary requests

This will be possible with the `getAuthenticatedRequest` method.

```php
// The server provides a way to get an authenticated API request for
// the service, using the access token; it returns an object conforming
// to Psr\Http\Message\RequestInterface.
$request = $server->getAuthenticatedRequest(
    'GET',
    'http://your.service/endpoint',
    $tokenCredentials
);

$client = $server->getHttpClient();

$response = $this->getHttpClient()->send($request);
```

**Comment by stevenmaguire** [comment]
@bencorlett I see you are circling this project again. What are your thoughts on moving along this refactor? What is left to accomplish?

**Comment by bencorlett** [comment]
Good question!

Let me pull the code down today and have a look, I really need to stop being such a slacker haha.

I’ll do that today :)

**Comment by shehi** [comment]
Any ETA on v2 folks?

**Comment by GrahamCampbell** [comment]
Any movement here?

**Comment by mikemand** [comment]
Any idea when this will be finished? Guzzle 4 is also EOL now, and v1 is still requiring Guzzle 3.

**Comment by bencorlett** [comment]
I'm back actively spending time on this project and I'm hoping to have the final refactoring done from the `feature/switch-to-phpspec` branch merged in within the week to release a beta of `v2`.

Following that, we just need to add implementations for more OAuth2 Servers and then we are good tor release.

**Comment by elazar** [comment]
@bencorlett Not to put too much more pressure on you, but a patch for a security vulnerability was included in Guzzle [6.2.1](https://github.com/guzzle/guzzle/blob/3f808fba627f2c5b69e2501217bf31af349c1427/CHANGELOG.md#621---2016-07-18). I'm assuming the vulnerability exists in all prior versions and that those versions won't receive patches. So, the sooner the switch to Guzzle 6 can happen, the better. 😄

**Comment by bencorlett** [comment]
Ooo, interesting. Could you please provide me with a link? I might see if I can also patch 1.x to run on a newer Guzzle as well without needing a major bump. Pretty sure we won't be modifying any public methods so it could be a patch version. 

Sent from my iPhone

Please excuse my brevity

**Comment by elazar** [comment]
@bencorlett Here's the site for the vulnerability: httpoxy.org. The Guzzle release notes I linked in my earlier comment include a related item.

File: composer.json [hunk 1]
```diff
@@ -1,46 +1,63 @@
...
+    "phploc-log": "./vendor/bin/phploc --log-csv build/loc.csv --progress --git-repository . src",
+    "phploc": "./vendor/bin/phploc src",
+    "phpspec": "./vendor/bin/phpspec run --no-interaction",
+    "behat": "./vendor/bin/behat --tags '@completed' --no-interaction",
+    "test": [
+      "@composer install",
+      "@phpcs",
+      "@phpcpd",
+      "@phploc",
+      "@phpspec",
```
Comment (stevenmaguire) [inline composer.json]: This test fails on the HHVM environment in Travis. Do we want to ignore HHVM? Do we want to find a new test solution?

File: composer.json [hunk 2]
```diff
@@ -1,46 +1,63 @@
...
+      "@behat"
```
Comment (stevenmaguire) [inline composer.json]: There are no integration tests in the project currently? Do we expect there to be any?

File: composer.json [hunk 3]
```diff
@@ -1,46 +1,63 @@
...
+    "phpcpd": "./vendor/bin/phpcpd src",
+    "phploc-log": "./vendor/bin/phploc --log-csv build/loc.csv --progress --git-repository . src",
+    "phploc": "./vendor/bin/phploc src",
+    "phpspec": "./vendor/bin/phpspec run --no-interaction",
+    "behat": "./vendor/bin/behat --tags '@completed' --no-interaction",
+    "test": [
+      "@composer install",
+      "@phpcs",
+      "@phpcpd",
+      "@phploc",
```
Comment (stevenmaguire) [inline composer.json]: The underlying package here requires at least PHP 5.6 and this package currently supports 5.5. Should we skip inclusion of this package?

File: src/Exception/ConfigurationException.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * Handles a temporary identifier mismatch.
+     *
+     * @return ConfigurationException
+     */
+    public static function temporaryIdentifierMismatch()
+    {
+        return new static(
+            'Temporary identifier passed back by server does not match that of stored temporary credentials. Potential man-in-the-middle.'
```
Comment (stevenmaguire) [inline src/Exception/ConfigurationException.php]: This line is too long, it will break code sniffer tests.

**Comment by stevenmaguire** [comment]
@bencorlett I am still interested in helping to get this project moved up and out to 2.0. It seems like the current test suite configuration is causing some issues with CI. I've added some comments and would like some input and/or direction from you on how we should proceed.

I am happy to get some more work done here :)

**Comment by stevenmaguire** [comment]
Yoo-hoo @bencorlett

**Comment by prisis** [comment]
Do you need some help or is this project dead?

**Comment by bencorlett** [comment]
Progress on the project has kind of screamed to a halt. I did speak with @philsturgeon about having @stevenmaguire or somebody else equally keen to take over the lead. I would be more than happy for this, just unsure what his movements are and whether he is interested :)

**Comment by philsturgeon** [comment]
We had this conversation over a year ago @bencorlett so yeah I'd say the swap should happen! 😅 

I'm not involved in the League anymore, so pester @frankdejonge or somebody to get this a new home.

**Comment by stevenmaguire** [comment]
I'm happy to take on more responsibility with this project. I'll need to get back up to speed a bit. I've been very involved in this last round of changes but it's been quite some time since my head has been in this space :)

**Comment by shehi** [comment]
Any update here? :) We have Guzzle 7 now with PHP 7.2+ support.

File: composer.json (hunk 1 above)
Comment (shehi) [reply composer.json:59]: I'd recommend ignoring HHVM at this age, since PHP 7 is as fast. PHP ecosystem has changed a lot since HHVM's inception.

**Comment by shehi** [comment]
Another question of mine is: who uses OAUTH-1 nowadays?! Haven't whole world moved to OAUTH-2 already?

**Comment by GrahamCampbell** [comment]
Twitter still uses oauth 1 only.

**Comment by shehi** [comment]
God damn it Twitter! :(

**Comment by shehi** [comment]
@GrahamCampbell , are you sure? https://developer.twitter.com/en/docs/authentication/oauth-2-0

**Comment by mfn** [comment]
Indeed, all the "real stuff" is still OAuth 1.0a => https://developer.twitter.com/en/docs/authentication/oauth-1-0a
(basically the most important stuff for serious Twitter interaction)

**Comment by shehi** [comment]
Yea. I know for sure almost noone will bother updating this package. Tech is too old to be of any value to anyone.

Maybe for the sake of moving forward, at Socialite side there should be new major version release with Oauth-1 and Twitter left behind. Otherwise everything will be stuck in past.

**Comment by shehi** [comment]
Can someone add me as a Member to this repo?

**Comment by shehi** [comment]
I need that to be able to stop Travis-CI runs which I know will fail. Waiting for them takes too long.

**Comment by shehi** [comment]
After spending 7-8 hours to refactor this codebase, I can say the following:

- This codebase has very critical typing issues. Many methods can return totally unrelated, multiple types (sometimes int|string, sometimes object|string, even void|string as well).
- For some reason it attempts to re-implement PHP's `http_build_query()` function and with errors at that.
- Even after I completely refactored it, adding `declare(strict_types=1)` in all files is completely impossible. One of the main methods, `sign()` method attempts to use non-strings (booleans, nulls and similar) as strings, url-decoding and re-encoding them back and forth, which completely breaks the outcome. I attempted to add fixes there, but it was too much hassle to go through, so no PHP declarations for the moment.

Overall, even though there is Version-2 branch half-done, I'd strongly suggest actually redoing that work on top of my refactoring, also adding `declare()` s to files.

**Pull request closed without merging**